File: src/nsl/statement/SwitchStatement.java
/*
 * SwitchStatement.java
 */

package nsl.statement;

import java.io.IOException;
import java.util.ArrayList;
import java.util.EnumSet;
import nsl.*;
import nsl.expression.*;

/**
 * Represents a switch statement.
 * @author Stuart
 */
public class SwitchStatement extends Statement
{
  private final Expression switchExpression;
  private final ArrayList<Statement> statementList;
  private final ArrayList<SwitchCaseStatement> casesList;

  /**
   * Class constructor.
   */
  public SwitchStatement()
  {
    if (!SectionInfo.in() && !FunctionInfo.in())
      throw new NslContextException(EnumSet.of(NslContext.Section, NslContext.Function), "switch");

    ScriptParser.tokenizer.matchOrDie('(');
    this.switchExpression = Expression.matchComplex();
    ScriptParser.tokenizer.matchOrDie(')');
    ScriptParser.tokenizer.matchOrDie('{');

    // Set non-null values so that the block statement can contain break statements.
    CodeInfo.getCurrent().setBreakLabel(RelativeJump.Zero);

    this.statementList = new ArrayList<Statement>();
    this.casesList = new ArrayList<SwitchCaseStatement>();

    // Get the statements including case statements.
    while (true)
    {
      Statement statement;
      if (ScriptParser.tokenizer.tokenIs("case"))
      {
        statement = new SwitchCaseStatement();
        this.casesList.add((SwitchCaseStatement)statement);
      }
      else
      {
        statement = Statement.match();
        if (statement == null)
          break;
      }

      this.statementList.add(statement);
    }

    // Check the last statement is a break statement.
    boolean noBreak = true;
    if (!this.statementList.isEmpty())
    {
      Statement last = this.statementList.get(this.statementList.size() - 1);
      if (last instanceof BlockStatement)
        last = ((BlockStatement)last).getLast();
      if (last instanceof BreakStatement)
        noBreak = false;
    }
    if (noBreak)
      throw new NslException("A \"switch\" statement must end with a \"break\" statement", true);

    CodeInfo.getCurrent().setBreakLabel(null);

    ScriptParser.tokenizer.matchOrDie('}');
  }

  /**
   * Assembles the source code.
   * @throws IOException
   */
  @Override
  public void assemble() throws IOException
  {
    // Do not assemble anything if there are no cases!
    if (this.casesList.isEmpty())
      return;

    // Give each case a label.
    for (SwitchCaseStatement statement : this.casesList)
      statement.setLabel(LabelList.getCurrent().getNext());

    Label gotoEnd = LabelList.getCurrent().getNext();
    Label gotoStart = LabelList.getCurrent().getNext();

    ScriptParser.writeLine("Goto " + gotoStart);
    
    Label parentBreak = CodeInfo.getCurrent().setBreakLabel(gotoEnd);

    for (Statement statement : this.statementList)
      statement.assemble();
    
    CodeInfo.getCurrent().setBreakLabel(parentBreak);

    gotoStart.write();

    Expression varOrSwitchExpression = AssembleExpression.getRegisterOrExpression(this.switchExpression);
    for (SwitchCaseStatement statement : this.casesList)
    {
      if (statement.getMatch().getType() == ExpressionType.Integer)
        ScriptParser.writeLine(String.format("IntCmp %s %s %s", varOrSwitchExpression, statement.getMatch(), statement.getLabel()));
      else
        ScriptParser.writeLine(String.format("StrCmp%s %s %s %s", statement.getMatch().getType() == ExpressionType.StringSpecial ? "S" : "", varOrSwitchExpression, statement.getMatch(), statement.getLabel()));
    }
    varOrSwitchExpression.setInUse(false);

    gotoEnd.write();
  }
}

File: src-dotnet/statement/SwitchStatement.cs
/*
 * SwitchStatement.java
 */
using Java.Io;
using Java.Util;
using Nsl;

namespace Nsl.Statement
{
    public class SwitchStatement : Statement
    {
        private readonly Expression switchExpression;
        private readonly List<Statement> statementList;
        private readonly List<SwitchCaseStatement> casesList;
        private SwitchDefaultCaseStatement defaultCase;
        /// <summary>
        /// Class constructor.
        /// </summary>
        public SwitchStatement()
        {
            if (!SectionInfo.In() && !FunctionInfo.In())
                throw new NslContextException(EnumSet.Of(NslContext.Section, NslContext.Function), "switch");
            int lineNo = ScriptParser.tokenizer.Lineno();
            ScriptParser.tokenizer.MatchOrDie('(');
            this.switchExpression = Expression.MatchComplex();
            ScriptParser.tokenizer.MatchOrDie(')');
            ScriptParser.tokenizer.MatchOrDie('{');

            // Set non-null values so that the block statement can contain break statements.
            CodeInfo.GetCurrent().SetBreakLabel(RelativeJump.Zero);
            this.statementList = new List<Statement>();
            this.casesList = new List<SwitchCaseStatement>();
            this.defaultCase = null;

            // Get the statements including case statements.
            while (true)
            {
                if (ScriptParser.tokenizer.Match("case"))
                {
                    Statement statement = new SwitchCaseStatement();
                    if (this.defaultCase != null)
                        throw new NslException("The \"default\" case in a \"switch\" statement must be the last case", true);
                    this.casesList.Add((SwitchCaseStatement)statement);
                    this.statementList.Add(statement);
                }
                else if (ScriptParser.tokenizer.Match("default"))
                {
                    this.defaultCase = new SwitchDefaultCaseStatement();
                    this.statementList.Add(this.defaultCase);
                }
                else
                {
                    Statement statement = Statement.Match();
                    if (statement == null)
                        break;
                    this.statementList.Add(statement);
                }
            }


            // No cases?
            if (this.casesList.IsEmpty())
                throw new NslException("A \"switch\" statement must have at least one \"case\" statement", true);

            // Validate switch cases for jump instructions.
            if (this.switchExpression is JumpExpression)
                ((JumpExpression)this.switchExpression).CheckSwitchCases(this.casesList, lineNo);

            // Check the last statement is a break statement.
            bool noBreak = true;
            if (!this.statementList.IsEmpty())
            {
                Statement last = this.statementList[this.statementList.Count - 1];
                if (last is BlockStatement)
                    last = ((BlockStatement)last).GetLast();
                if (last is BreakStatement)
                    noBreak = false;
            }

            if (noBreak)
                throw new NslException("A \"switch\" statement must end with a \"break\" statement", true);
            CodeInfo.GetCurrent().SetBreakLabel(null);
            ScriptParser.tokenizer.MatchOrDie('}');
        }

        /// <summary>
        /// Class constructor.
        /// </summary>
        // Set non-null values so that the block statement can contain break statements.
        // Get the statements including case statements.
        // No cases?
        // Validate switch cases for jump instructions.
        // Check the last statement is a break statement.
        public override void Assemble()
        {

            // Do not assemble anything if there are no cases!
            if (this.casesList.IsEmpty())
                return;

            // Give each case a label.
            foreach (SwitchCaseStatement statement in this.casesList)
                statement.SetLabel(LabelList.GetCurrent().GetNext());
            if (this.defaultCase != null)
                this.defaultCase.SetLabel(LabelList.GetCurrent().GetNext());
            Label gotoEnd = LabelList.GetCurrent().GetNext();
            Label gotoStart = LabelList.GetCurrent().GetNext();

            // Go to the jump table which is assembled after the switch case labels and
            // statements.
            ScriptParser.WriteLine("Goto " + gotoStart);

            // Using "break;" inside a switch jumps to the end.
            Label parentBreak = CodeInfo.GetCurrent().SetBreakLabel(gotoEnd);

            // Assemble all the statements inside the switch { }. This includes the
            // case labels.
            foreach (Statement statement in this.statementList)
                statement.Assemble();

            // Restore the parent break label.
            CodeInfo.GetCurrent().SetBreakLabel(parentBreak);

            // Label at the top of the jump table.
            gotoStart.Write();

            // Jump instructions can jump directly to the case labels.
            if (this.switchExpression is JumpExpression)
            {
                ((JumpExpression)this.switchExpression).Assemble(this.casesList);
            }
            else

            // Other expressions we just assemble them if required and compare their
            // result.
            {
                Expression varOrSwitchExpression = AssembleExpression.GetRegisterOrExpression(this.switchExpression);
                foreach (SwitchCaseStatement caseStatement in this.casesList)
                {

                    // Type is an integer; use IntCmp.
                    if (caseStatement.GetMatch().GetType().Equals(ExpressionType.Integer))
                    {
                        ScriptParser.WriteLine(String.Format("IntCmp %s %s %s", varOrSwitchExpression, caseStatement.GetMatch(), caseStatement.GetLabel()));
                    }
                    else

                    // Type is a string; use StrCmp or StrCmpS (if special `` quotes were
                    // used).
                    {
                        ScriptParser.WriteLine(String.Format("StrCmp%s %s %s %s", caseStatement.GetMatch().GetType().Equals(ExpressionType.StringSpecial) ? "S" : "", varOrSwitchExpression, caseStatement.GetMatch(), caseStatement.GetLabel()));
                    }
                }

                varOrSwitchExpression.SetInUse(false);
            }


            // Default case jump.
            if (this.defaultCase != null)
                ScriptParser.WriteLine("Goto " + this.defaultCase.GetLabel());
            gotoEnd.Write();
        }
    }
}